### ISSUE TITLE
False positive on Starlette middleware registration

### ISSUE THREAD

**Issue by Asaf31214** [astral-sh/ty#1635]
### Summary

Register any FastAPI middleware, custom or built-in:
``` python
app = FastAPI()
app.add_middleware(NoStoreAuthMiddleware) # custom, subclass of BaseHTTPMiddleware
app.add_middleware(RateLimitMiddleware) # custom, subclass of BaseHTTPMiddleware
app.add_middleware( 
    CORSMiddleware, # built-in middleware, from fastapi.middleware.cors
    allow_origins=["*"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
) 
```
run:
`ty check`, without any specified rules at `pyproject.toml`

No issue until ty@0.0.1a25 (expected result)
But for versions ty@0.0.1a26 and ty@0.0.1a27 (current latest):
```
error[invalid-argument-type]: Argument to bound method `add_middleware` is incorrect
  --> main.py:31:20
   |
30 | app = FastAPI(default_response_class=ORJSONResponse, lifespan=lifespan)
31 | app.add_middleware(NoStoreAuthMiddleware)
   |                    ^^^^^^^^^^^^^^^^^^^^^ Expected `_MiddlewareFactory[Unknown]`, found `<class 'NoStoreAuthMiddleware'>`
32 | app.add_middleware(RateLimitMiddleware)
33 | app.add_middleware(ExceptionHandlerMiddleware)
   |
info: Method defined here
   --> .venv/lib/python3.14/site-packages/starlette/applications.py:118:9
    |
116 |         self.router.host(host, app=app, name=name)  # pragma: no cover
117 |
118 |     def add_middleware(
    |         ^^^^^^^^^^^^^^
119 |         self,
120 |         middleware_class: _MiddlewareFactory[P],
    |         --------------------------------------- Parameter declared here
121 |         *args: P.args,
122 |         **kwargs: P.kwargs,
    |
info: rule `invalid-argument-type` is enabled by default
```
same error for all 3 middlewares registered here.

this is also mentioned at issue #1564  by @dmytro-GL [here](https://github.com/astral-sh/ty/issues/1564#issuecomment-3548995805)

### Version

ty 0.0.1-alpha.27

**Comment by carljm** [comment]
Thanks for the report! It looks like `_MiddlewareFactory` is a protocol which is generic over a `ParamSpec`, so I think we should revisit this once #157 is done.

**Comment by carljm** [comment]
Putting this in beta milestone as I think we should ensure it's either fixed or suppressed in some way for the beta.

**Comment by asaf-kanlipicak** [comment]
Update: same behaviour on 0.0.1-alpha.28

**Comment by dhruvmanila** [comment]
For context, this is same as https://github.com/astral-sh/ty/issues/157#issuecomment-3554317390 which is a self-contained example to reproduce this. This possibly requires #1714.

**Comment by carljm** [comment]
Didn't get this into the beta, but it's a high priority follow-up.

**Comment by thejcannon** [comment]
(as far as examples goes, it doesn't get much smaller than this)
```
#!/usr/bin/env -S uv run --with starlette==0.50.0,ty==0.0.2 ty check --

from starlette_context.middleware import RawContextMiddleware
from starlette.middleware import Middleware

Middleware(RawContextMiddleware)
```

**Comment by condemil** [comment]
Another minimal reproducible example:


```python
from starlette.middleware import Middleware
from starlette.middleware.httpsredirect import HTTPSRedirectMiddleware

# ty(invalid-argument-type): Argument to bound method `__init__` is incorrect: Expected `_MiddlewareFactory[(...)]`, found `<class 'HTTPSRedirectMiddleware'>`
middleware = Middleware(HTTPSRedirectMiddleware)
```

**Comment by Kludex** [comment]
When this is fixed, I'm open to add `ty` to the `starlette`s pipeline.

**Comment by carljm** [comment]
https://github.com/astral-sh/ruff/pull/21902 wasn't sufficient to fix this -- I think there's still an issue here related to ParamSpec.

**Comment by Kludex** [comment]
Thanks @zanieb and @dhruvmanila. ❤️
